### ISSUE TITLE
Override matmul gradient function to support random feedback gradients

### ISSUE THREAD

**Issue by benlansdell** [benlansdell/synthfeedback#10]

**Comment by benlansdell** [comment]
Some reading:
https://stackoverflow.com/questions/43839431/tensorflow-how-to-replace-or-modify-gradient

https://www.tensorflow.org/extend/adding_an_op

https://medium.com/@Sirena/override-tensorflow-backward-propagation-1a7e2bb3f516

**Comment by benlansdell** [comment]
Actually, easiest solution is to take suggestion from: https://stackoverflow.com/questions/43839431/tensorflow-how-to-replace-or-modify-gradient

Use the hack:

```y = b(x) + tf.stop_gradient(f(x) - b(x))```

This acts like f(x) on forward pass, and like b(x) on backward pass.

**Comment by benlansdell** [comment]
This passes a unit test. Need to change implementation of feedback alignment to make use of it....

**Comment by benlansdell** [comment]
This hack actually didn't work. If computing L = Ax, the above hack works for computing dL/dx, but not dL/dA. It just sets this to zero, which we don't want. 

The solution is instead to use the example here:

https://github.com/davidstutz/tensorflow-cpp-op-example

In particular, the .py file in this repo has an example that can be modified.

**Comment by benlansdell** [comment]
See also: http://davidstutz.de/implementing-tensorflow-operations-in-c-including-gradients/

**Comment by benlansdell** [comment]
This example is also useful:
https://stackoverflow.com/questions/39048984/tensorflow-how-to-write-op-with-gradient-in-python
